### ISSUE TITLE
Implement `jj sync` command

### ISSUE THREAD

**Issue by dbarnett** [martinvonz/jj#1039]
jj should have a command to sync all of my branches (or some subset) with a remote via merge or rebase, in a single operation.

Since that could cause a lot of big changes at once, it'd be nice to show a preview of changes that will be performed and prompt to confirm and/or output a very clear summary of what all changed from that sync operation.

**Comment by dbarnett** [comment]
I could also imagine some related curation tools for e.g. moving a local branch to "draft mode" to not be published or synced yet.

**Comment by dbarnett** [comment]
This is also one of the areas jj was dinged on in the comparison table at https://blog.waleedkhan.name/git-ui-features/#data.

**Comment by MForster** [comment]
IMHO `sync` is ambiguous about the direction. Does it only pull or also push?

**Comment by PhilipMetzger** [comment]
I think `jj sync` should only pull, like its [Perforce predecessor](https://www.perforce.com/manuals/cmdref/Content/CmdRef/p4_sync.html). If that isn't clear enough, we can alias the command for Git users to `jj pull`

**Comment by martinvonz** [comment]
> IMHO `sync` is ambiguous about the direction. Does it only pull or also push?

Hmm, that's a good point. On the other hand, I don't think we've had any confusion about that at Google with our `hg sync` command, which does pull+rebase. I'm not sure what a better name would be.

> we can alias the command for Git users to `jj pull`

I think Git's use of "pull" for "fetch+rebase/merge" is quite confusing. I prefer hg's use of `hg pull` for just pulling (like `git fetch`). That is more symmetric with `hg/git push`.

**Comment by MForster** [comment]
> I don't think we've had any confusion about that at Google with our `hg sync` command

That's not representative, because ~everybody at Google has used Perforce/Piper.

FWIW, VSCode has a "Sync" button that pulls and pushes...

**Comment by martinvonz** [comment]
> That's not representative, because ~everybody at Google has used Perforce/Piper.

True. (I have practically not used it, but I still have run `p4 sync` several times.)

> FWIW, VSCode has a "Sync" button that pulls and pushes...

Oh, that's a good data point!

We'd still have to think of a better name then... There's `hg fetch` that does this. I don't know if they named it that as a joke, given that it means that `hg pull` == `git fetch` and `hg fetch` == `git pull`. So I don't really think we should call it `jj fetch`.

**Comment by PhilipMetzger** [comment]
IMHO, `jj sync` should be good enough. We then should mention, that it's inspired by Perforce's `p4 sync`. If we then want to accommodate Mercurial and Git users, we can alias it to their equivalent. 
I think we should approach Editor integration with our own tools and assumptions and disregard the current state for Git and Mercurial.

**Comment by dbarnett** [comment]
Graphite also uses "sync" to refer to this same kind of "pull-and-restack" operation: https://docs.graphite.dev/guides/graphite-cli/syncing-and-resolving-conflicts

Do we envision jj ever would support that kind of bidirectional "pull + push"? If so, maybe we have a command called sync that defaults to `--in` mode with eventual support for `--in-out` mode, or something like that? (Hard for me to imagine wanting such a shotgun command though, personally.)

**Comment by Ralith** [comment]
> That's not representative, because ~everybody at Google has used Perforce/Piper.

I skipped Piper and went straight to Fig, and found `hg sync` to be plenty intuitive. Agreed that doing both at once seems weird and error-prone.

**Comment by ilyagr** [comment]
Now that https://github.com/martinvonz/jj/pull/1288 is (hopefully) in, you can get part of the way to `jj sync` with a command like `jj rebase -s oldmain+~:main -d main -L`. This is, of course, not particularly pretty and requires finding the `oldmain` commit.

Another, simpler, alternative if you're OK with rebasing everything not on main: `jj rebase -s roots(main..) -d main`.

**Comment by PhilipMetzger** [comment]
We discussed `jj sync` multiple times in the recent months in Discord as there is one big issue to solve: backend independence. 

**Summary:**
Ideally `jj sync` should be a un-namespaced command (without a backend prefix) but this makes it difficult to determine which implementation to use. @martinvonz's suggesion in an internal chat and on Discord was to make the command forge specific, e.g `jj git sync`, `jj gerrit sync` or `jj piper sync` and then allow the unnamespaced sync command to wrap it. This means that for the most users, `jj sync` would be `jj git sync`. Another option is to make that a buiiltin alias for each backend. 

While I was initially sceptical of the namespaced approach, I now think it is the right solution for this problem. The only issue I see, if we take the forge-specific approach, is that some layering issue may pop up. But if we implement something like `jj submit` or `jj mail`, a forge trait is needed anyway.

**Comment by essiene** [comment]
I put together a mini proposal at https://tinyurl.com/jj-git-sync-dd

I'd like to get some thoughts especially around the semantics of the rebase itself. Also, if I missed some edge cases that need special handling.

Thanks

**Comment by PhilipMetzger** [comment]
I'm going to take a look at it tomorrow, to see if it matches the command we discussed in Discord. From a quick look, it's currently missing a way to create a merge commit like `jj git sync --merge <Bookmark0...N>` as not everyone is using the rebase workflow.

**Comment by joyously** [comment]
Are the `git` commands supposed to mirror Git? Like is there a `pull` and a `fetch`? Are both needed, or is this a way to do it better?

**Comment by essiene** [comment]
> I'm going to take a look at it tomorrow, to see if it matches the command we discussed in Discord. From a quick look, it's currently missing a way to create a merge commit like `jj git sync --merge <Bookmark0...N>` as not everyone is using the rebase workflow.

Could you summarize what the behavior is supposed to be in this case? I'm a rebase-tionist (ha!) so my thinking around this is probably wrong. Would this do something like:
  - fetch
  - merge specified bookmarks into `main` ?

**Comment by arxanas** [comment]
> I'll answer in the DDOC, as it is the more appropriate place for the ensuing discussion.

Link to that discussion: https://docs.google.com/document/d/1K4j1xQPFxtINhyOyKuCMTZ_iNrw8g-5_giBFMGoLjaI/edit?pli=1&disco=AAABU_7EcdY

**Comment by hiasr** [comment]
@essiene Would love to see this come through! Is there anywhere I can help?

**Comment by hotsphink** [comment]
Drive-by comment: personally, I think `jj git pullup` makes sense as a name. You can read it as either "pull+update" or "pull my local repo up to the state of the remote repo", and the direction is in the name.

The name `jj git pushup` is of course available as a future command.

**Comment by ahal** [comment]
Here's an alias that accomplishes a hacky version of this in the meantime:
```toml
[aliases]
pull = ["util", "exec", "--", "bash", "-c", """
if [ -z "$1" ]; then
    echo "missing bookmark"
    exit 1
fi
set -euo pipefail
jj git fetch -b $1
jj rebase -s "at_operation(@-, $1+) & mutable()" -d $1
""", ""]
```

Use it like `jj pull main`

**Comment by dbarnett** [comment]
I tried that recipe but it somehow didn't work and left me with a bunch of unexpected divergent changes. Still haven't had a chance to investigate why. 

Do we have a good robust recipe to recommend for that? What I'd like such a command to handle is: 

1. Note the current local `trunk()` change ID, call it "oldtrunk"
2. Fetch remote
3. For any of the local changes in `oldtrunk+ ~ immutable()` not logically present in the remote, rebase onto the new trunk
4. For any that ARE logically present (e.g., I merged a PR including those changes), abandon… UNLESS they have descendants that need special handling

Does that sound like a reasonable set of behaviors for this sync / "pullup" thing to do?

**Comment by Dynge** [comment]
I think your divergent changes can be fixed by adding `--skip-emptied` to the rebase.

This should auto abandon changes that become empty when rebased (they would become empty if all changes are already included in the new trunk.)

**Comment by dbarnett** [comment]
Another issue I just ran into with rebasing... is there a proper way to reparent one of multiple parents for a change?

```text
@    zozquuxv 2025-12-01 10:23:48 c3dd3da3
├─╮  Merging otherbranch and main
│ ○  yuzmlwzy 2025-12-01 10:23:29 otherbranch ee7f2523
├─╯  
○  mmpmpkzn 2025-12-01 10:23:04 main git_head() 97d906c5
│  
```

How would I swap in a new `main` and rebase changes from the old main to keep the same structure but move that one parent forward? Can I do that with a single command?

(Sorry to go a little deep in a corner case here, I'm just wondering if there's one straightforward general recipe to pull/update and hoping to understand if special cases like this can be generalized…)

**Comment by martinvonz** [comment]
@dbarnett: I think you're looking for `jj simplify-parents`

**Comment by dbarnett** [comment]
Not sure how to use `simplify-parents` for the case of rebasing X onto `X- ~ Y | Y2` (all X's existing parents except Y, plus Y's new replacement Y2). Did you mean to do some rebasing first and then simplify?

**Comment by SeerLite** [comment]
@dbarnett A bit late but would `jj rebase -s main@git -o main@origin` work?

**Comment by dbarnett** [comment]
@SeerLite no, `main@git` is handy to use but it's the same limitation that  `-o` or `-d` will replace all parents instead of just the one. 

For this case I've started introducing a empty placeholder before the rebase and abandoning it after, like: 
```shell
jj new -A stale-local-change
jj rebase -s @ -d main
jj abandon @ oldmain+::stale-local-change
```
It's a little roundabout but gets me one handle to grip and pivot the stale parent relationships by without affecting the others.

**Comment by chrish42** [comment]
For me, the lack of this command is the biggest blocker to recommending jj to more people in my team: jj + Github has as special, unintuitive little dance needed for squash and rebase workflows, which are very common. If that part was also handled smartly by some kind of `jj git` command, then it would be a no-brainer to recommend jj to folks that are collaborating on team git repos.

**Comment by LoganDark** [comment]
I've been having an issue where `jj git fetch` keeps creating divergent upstream changes instead of rebasing my local changes on top of the new upstream versions. Right now the only solution is to write a script that crawls the entire commit graph and runs hundreds of manual rebases, which is fragile and slow. Is the topic of this issue in any way related to that?

**Comment by jwhpryor** [comment]
Is this for feature branches? I think you would see a divergence if someone re-uploaded the branch. I use `jj git fetch` and never have issues with divergences, but I would expect this to happen if a feature branch went sideways.
